### PR TITLE
Migrate to showing prime chaches progress

### PR DESCRIPTION
**Pull request by Veetaha**
This is a part of #4074

### CODE REVIEW

**Comment by Veetaha** [comment]
Hmm, I noticed there was a bug in progress reporting.
It could be so that we did create the progress resource but reported it's end (i.e. this is an analogue of free of a dangling pointer in LSP)
See this for example (I added logs to the impl):
```diff
fn send_startup_progress(sender: &Sender<Message>, loop_state: &mut LoopState) {
    let total: usize = loop_state.roots_total;
    let prev = loop_state.roots_progress_reported;
    let progress = loop_state.roots_scanned;
    loop_state.roots_progress_reported = Some(progress);

+    eprintln!("SEND STARTUP PROGRESS");

    match (prev, loop_state.workspace_loaded) {
        (None, false) => {
+            eprintln!("SEND STARTUP PROGRESS: BEGINNING");
            let work_done_progress_create = request_new::<req::WorkDoneProgressCreate>(
                loop_state.next_request_id(),
                WorkDoneProgressCreateParams {
                    token: req::ProgressToken::String("rustAnalyzer/startup".into()),
                },
            );
            sender.send(work_done_progress_create.into()).unwrap();
            send_startup_progress_notif(
                sender,
                WorkDoneProgress::Begin(WorkDoneProgressBegin {
                    title: "rust-analyzer".into(),
                    cancellable: None,
                    message: Some(format!("{}/{} packages", progress, total)),
                    percentage: Some(100.0 * progress as f64 / total as f64),
                }),
            );
        }
        (Some(prev), false) if progress != prev => {
+            eprintln!("SEND STARTUP PROGRESS: REPORTING {}/{}", progress, total);
            send_startup_progress_notif(
                sender,
                WorkDoneProgress::Report(WorkDoneProgressReport {
                    cancellable: None,
                    message: Some(format!("{}/{} packages", progress, total)),
                    percentage: Some(100.0 * progress as f64 / total as f64),
                }),
            )
        },
        (_, true) => {
+            eprintln!("SEND STARTUP PROGRESS: END: {}/{}", progress, total);
            send_startup_progress_notif(
                sender,
                WorkDoneProgress::End(WorkDoneProgressEnd {
                    message: Some(format!("rust-analyzer loaded, {} packages", progress)),
                }),
            )
         },
        _ => {}
    }

    fn send_startup_progress_notif(sender: &Sender<Message>, work_done_progress: WorkDoneProgress) {
        let notif = notification_new::<req::Progress>(req::ProgressParams {
            token: req::ProgressToken::String("rustAnalyzer/startup".into()),
            value: req::ProgressParamsValue::WorkDone(work_done_progress),
        });
        sender.send(notif.into()).unwrap();
    }
}
```
<details>
<summary>Log output</summary>

```

~/dev/rust-analyzer (master) $ RUST_LOG=rust_analyzer=debug RA_PROFILE=* cargo test -p rust-analyzer --test heavy_tests -- --nocapture test_missing_module_code_action_in_json_project --exact
   Compiling stdx v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/stdx)
   Compiling ra_syntax v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/ra_syntax)
   Compiling ra_cfg v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/ra_cfg)
   Compiling ra_mbe v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/ra_mbe)
   Compiling ra_fmt v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/ra_fmt)
   Compiling ra_db v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/ra_db)
   Compiling ra_proc_macro_srv v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/ra_proc_macro_srv)
   Compiling ra_hir_expand v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/ra_hir_expand)
   Compiling ra_project_model v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/ra_project_model)
   Compiling ra_hir_def v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/ra_hir_def)
   Compiling ra_hir_ty v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/ra_hir_ty)
   Compiling ra_hir v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/ra_hir)
   Compiling ra_ide_db v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/ra_ide_db)
   Compiling ra_assists v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/ra_assists)
   Compiling ra_ide v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/ra_ide)
   Compiling rust-analyzer v0.1.0 (/home/veetaha/dev/rust-analyzer/crates/rust-analyzer)
    Finished test [unoptimized] target(s) in 38.23s
     Running target/debug/deps/heavy_tests-ba0c74fdf8d9c385

running 1 test
[INFO  rust_analyzer::main_loop] initial config: Config {
        client_caps: ClientCapsConfig {
            location_link: true,
            line_folding_only: false,
            hierarchical_symbols: false,
            code_action_literals: true,
        },
        with_sysroot: false,
        publish_diagnostics: true,
        lru_capacity: None,
        proc_macro_srv: None,
        files: FilesConfig {
            watcher: Notify,
            exclude: [],
        },
        notifications: NotificationsConfig {
            cargo_toml_not_found: true,
        },
        cargo: CargoConfig {
            no_default_features: false,
            all_features: true,
            features: [],
            load_out_dirs_from_check: false,
            target: None,
        },
        rustfmt: Rustfmt {
            extra_args: [],
        },
        check: Some(
            CargoCommand {
                command: "check",
                all_targets: true,
                all_features: true,
                extra_args: [],
            },
        ),
        inlay_hints: InlayHintsConfig {
            type_hints: true,
            parameter_hints: true,
            chaining_hints: true,
            max_length: None,
        },
        completion: CompletionConfig {
            enable_postfix_completions: true,
            add_call_parenthesis: true,
            add_call_argument_snippets: true,
            snippet_cap: Some(
                SnippetCap {
                    _private: (),
                },
            ),
        },
        call_info_full: true,
    }
[WARN  rust_analyzer::world] Cargo check watching only supported for cargo workspaces, disabling
[INFO  rust_analyzer::main_loop] server initialized, serving requests
[INFO  rust_analyzer::main_loop] loop turn = VfsTask { ... }
SEND STARTUP PROGRESS
SEND STARTUP PROGRESS: END: 1/1
[INFO  rust_analyzer::main_loop] loop turn = Notification { method: "textDocument/didOpen" }
[INFO  rust_analyzer::main_loop] loop turn = Notification { method: "textDocument/didOpen" }
[INFO  rust_analyzer::main_loop] loop turn = Request(Request { id: RequestId(U64(1)), method: "textDocument/codeAction", params: Object({"context": Object({"diagnostics": Array([])}), "range": Object({"end": Object({"character": Number(7), "line": Number(0)}), "start": Object({"character": Number(4), "line": Number(0)})}), "textDocument": Object({"uri": String("file:///tmp/.tmpZikugY/src/lib.rs")})}) })
[INFO  rust_analyzer::main_loop] loop turn = Diagnostic(SetNative(FileId(0), [Diagnostic { range: Range { start: Position { line: 0, character: 0 }, end: Position { line: 0, character: 8 } }, severity: Some(Error), code: None, source: Some("rust-analyzer"), message: "unresolved module", related_information: None, tags: None }]))
[INFO  rust_analyzer::main_loop] loop turn = Response { id: RequestId(U64(1)), error: None }
[INFO  rust_analyzer::main_loop] handled req#1 in 8.801716ms
[INFO  rust_analyzer::main_loop] loop turn = Request(Request { id: RequestId(U64(2)), method: "textDocument/codeAction", params: Object({"context": Object({"diagnostics": Array([])}), "range": Object({"end": Object({"character": Number(7), "line": Number(2)}), "start": Object({"character": Number(4), "line": Number(2)})}), "textDocument": Object({"uri": String("file:///tmp/.tmpZikugY/src/lib.rs")})}) })
[INFO  rust_analyzer::main_loop] loop turn = Response { id: RequestId(U64(2)), error: None }
[INFO  rust_analyzer::main_loop] handled req#2 in 5.128107ms
[INFO  rust_analyzer::main_loop] waiting for tasks to finish...
[INFO  rust_analyzer::main_loop] ...tasks have finished
[INFO  rust_analyzer::main_loop] joining threadpool...
[INFO  rust_analyzer::main_loop] ...threadpool has finished
test test_missing_module_code_action_in_json_project ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 10 filtered out

~/dev/rust-analyzer (master) $ 
```
</details>

As you can see there is no `SEND STARTUP PROGRESS: BEGINNING` in the log output, so we freed the resource without creating it...

In my PR, we always create the progress resource right away at the beginning of the main loop so this should be fixed.

File: crates/rust-analyzer/src/main_loop.rs [hunk 1]
```diff
@@ -782,55 +805,12 @@ fn on_diagnostic_task(task: DiagnosticTask, msg_sender: &Sender<Message>, state:
...
-            send_startup_progress_notif(
-                sender,
-                WorkDoneProgress::Begin(WorkDoneProgressBegin {
-                    title: "rust-analyzer".into(),
-                    cancellable: None,
-                    message: Some(format!("{}/{} packages", progress, total)),
-                    percentage: Some(100.0 * progress as f64 / total as f64),
-                }),
-            );
+fn send_workspace_analisys_progress(loop_state: &mut LoopState) {
```
Comment (cynecx) [inline crates/rust-analyzer/src/main_loop.rs:808]: typo: analysis

**Comment by kjeremy** [comment]
I would like to see #4403 merged before this so we can handle clients that do not support progress.

File: crates/rust-analyzer/src/main_loop.rs (hunk 1 above)
Comment (Veetaha) [reply crates/rust-analyzer/src/main_loop.rs]: Thanks, fixed a1b6914e6425a2e78d7d298f02f0b63e73087bc0

File: crates/rust-analyzer/src/main_loop.rs [hunk 2]
```diff
@@ -443,7 +451,22 @@ fn loop_turn(
...
+            let mut progress = PrimeCachesProgressNotifier::begin(
+                connection.sender.clone(),
+                loop_state.next_request_id(),
+                total,
+            );
+
+            move || {
+                snap.analysis()
+                    .prime_caches(subs, move |i| {
+                        progress.report(i + 1);
```
Comment (matklad) [inline crates/rust-analyzer/src/main_loop.rs:466]: Note that we will initiate a new progress after every change, which is probably is not the right user experience. The `i` itself is also pretty misleading -- the number of open files (which is typically `1`) doesn't really have a relationship to the amount of work we need to do.

**Comment by kjeremy** [comment]
@Veetaha #4403 has merged.

**Comment by matklad** [comment]
ping :-) 

Though, my gut feeling is that https://github.com/rust-analyzer/rust-analyzer/pull/4407#discussion_r423184183 is a pretty significant problem, and that it would be hard to fix. Even conceptually, it's hard to show percentage things when you do stuff lazily. We can't know how many crates do we need to process upfront. 

it would be cool to maybe emit "Start/Finished/Cancelled" events for crate def map query specifically though...

**Comment by Veetaha** [comment]
@matklad in fact I have doubts that the current PR design is what is good, it'd be cool to see your feedback on some design thoughts I had for progress reporting. 
I didn't make progress on this PR since I ended up making our `next_request_id` changed from `u64` to `Arc<AtomicU64>` which now I think is not that bad.

The problem I tried to also solve is that our current cargo check progress reports don't actually fully abide by the lsp protocol (or at least I saw some bugs/flaws in it) and we do some custom handling of this on the frontend.

Well, to give a small peek on my design thoughts, here are some high levels:
```rust
//! Instrumentation for progress reporting.

#[cfg(feature = "lsp")]
pub mod lsp; // already written impls of Progress and ProgressProvider for LSP servers

// This will be used by flycheck
pub mod indeterminate {
    pub trait Progress {
        fn report(&mut self, message: String);
    }

    pub trait ProgressProvider {
        type Progress: Progress;

        fn begin(&mut self, message: String) -> Self::Progress;
    }
}

// This will be used for package analysis reports
pub mod determinate {
    pub struct IsDone(pub bool);

    /// Represents an interface to report progress of some runnings task.
    /// It is abstracted form the underlying logic for propagating the progress
    /// to ui or anywhere else.
    /// Internally records the state and returns `IsDone(true)` 
    /// when the progress has reached `total` or more. 
    ///
    /// The progress is a resource that needs to be released and only once thus
    /// it must provide a custom `Drop` for this purpose.
    pub trait Progress { // TODO: require Drop as a supertrait?
        fn report(&mut self, processed: usize) -> IsDone;
    }

    pub trait ProgressProvider {
        type Progress: Progress;

        /// Creates and registers the progress instace of `determinate::Progress`.
        /// It must be deregistered on its `drop()`.
        fn begin(&mut self, initial: usize, total: usize) -> Self::Progress;
    }
}
```
This design requires `ProgressProvider` to own a reference to `next_request_id`, hence `Arc<AtomicU64>`.
I could certainly evolve this idea (in fact, surprisingly, it is still in my `git stash`)

**Comment by matklad** [comment]
>The problem I tried to also solve is that our current cargo check progress reports don't actually fully abide by the lsp protocol

So this PR does two things, right?

1. Changes the actual LSP messages to be exactly like LSP protocol, so that we don't need custom coding on the client
2. Changes the "meaning" of progress, switching it from just loading crates to loading and processing crates? 

Let's scale it to just the first bullet first. 

In terms of design, I think it's better if we avoid providing generic infrastructure with traits and such here. I think it's better if we just track everything ourselves in the event loop. I feel "every process tick implies a turn of event loop" is an easier to debug design, than something based on Arced callbacks.

**Comment by Veetaha** [comment]
Well, initially the idea was just to show some status to the user when nothings seems to happen (highlighting doesn't show up, goto defs don't work, etc..). After having extracted `progress.rs` file with utils for reporting the progress in LSP server I thought I could go further by adopting these utils in flycheck crate. Unfortunatelly, I stumbled with `Arc<AtomicU64>` and abandoned the idea...

Now that you have mentioned processing everything on the event loop, I think I can avoid `Arcs` by having a `Sender` which sends progress notifications to the `main_loop` where in turn it sends a notification to the IDE client (i.e. the same design we have with `flycheck` now, but also applied to package analysis reports and all other long-running things like crate def map (as you have mentioned).

I tried doing this with traits, such that we have something consistent to be used across the codebase without the dependency on `lsp` (for example if we go for showing the progress of inital `cargo check` that is run to get `outdirs`, we could pass something like `impl ProgressProvider` so that `ra_project_model` can call `begin(total) -> ProgressToken` and then continuously `token.report(crate)` without the knowledge of how the report works under the hood).

**Comment by lnicola** [comment]
```rust
// This will be used by flycheck
pub mod indeterminate {
    pub trait ProgressProvider {
        fn begin(&mut self, message: String) -> Self::Progress;
    }
}
```

I'm pretty sure you didn't want to know this, but that's actually `complete`, not `begin`. We only see a crate _after_ it was checked.

**Pull request closed without merging**